### ISSUE TITLE
Segmentation loss

### ISSUE THREAD

**Issue by peterneher** [facebookarchive/caffe2#696]
Hi everyone,

is there a loss for semantic segmentation? E.g. dice loss or voxel-wise crossentropy?

Cheers,
Peter

**Comment by akyrola** [comment]
You can use SigmoidCrossEntropyWithLogits?

**Comment by futurely** [comment]
Based on the example
[`mask_loss = mask_lw * tf.nn.sigmoid_cross_entropy_with_logits(labels=mask_targets, logits=masks)`](
https://github.com/CharlesShang/FastMaskRCNN/blob/a84dac4bfcd6d4928da5417de04b862e18ad03f4/libs/nets/pyramid_network.py#L482), you can use [`mask_loss = model.net.SigmoidCrossEntropyWithLogits([masks, mask_targets], ['mask_loss'])`](
https://github.com/caffe2/caffe2/blob/f27fdbad55946feeba99b7c8c1e0ff899cfaf0b8/caffe2/python/operator_test/cross_entropy_ops_test.py#L63) or [`mask_loss = model.BatchSigmoidCrossEntropyLoss(input_record)`](https://github.com/caffe2/caffe2/blob/216fa5baffcb8b9e50cb6458a8764f14e58e414f/caffe2/python/layers_test.py#L184).

**Comment by peterneher** [comment]
Great, I will check that out. Thank you!

**Comment by peterneher** [comment]
The SigmoidCrossEntropyWithLogits expects a tensor as input that was passed trough a logit, right?
Unfortunately the gradient of the logit op is not implemented, or am I missing something?

**Comment by futurely** [comment]
SigmoidCrossEntropyWithLogits expects both the [predicted](https://github.com/CharlesShang/FastMaskRCNN/blob/a84dac4bfcd6d4928da5417de04b862e18ad03f4/libs/nets/pyramid_network.py#L311) and the ground truth mask tensors.

The details of the gradient op are in the [implemenation](https://github.com/caffe2/caffe2/search?utf8=%E2%9C%93&q=SigmoidCrossEntropyWithLogitsGradient&type=).

**Comment by peterneher** [comment]
Of course it also needs the ground truth as input. My point is that it the logit op is not fully implemented:
https://github.com/caffe2/caffe2/blob/d1fb979f37840ef42714f6f8fb1ee0028d127cf1/caffe2/operators/logit_op.cc#L47

**Comment by futurely** [comment]
`logits` is simply a [tensor](https://github.com/caffe2/caffe2/blob/b7289afd01606b963865ba9c5f96e284c0ec1069/caffe2/operators/cross_entropy_op.cc#L51). SigmoidCrossEntropyWithLogits does not use the Logits op.
https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits

**Comment by peterneher** [comment]
Thank you. I was confused by the name "logits" for an arbitrary input.
